src/Symfony/Component/Scheduler/Scheduler.php

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+final class Scheduler implements SchedulerInterface
+{
+    /**
+     * @var \DateTimeZone
+     */
+    private $timezone;
+
+    /**
+     * @var EventDispatcherInterface|null
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var TaskInterface[]
+     */
+    private $tasks = [];
+
+    /**
+     * @var MessageBusInterface|null
+     */
+    private $bus;
+
+    /**
+     * @var TaskPoolInterface[]
+     */
+    private $pools;
+
+    /**
+     * @param string|\DateTimeZone $timezone
+     */
+    public function __construct($timezone, array $pools = [], EventDispatcherInterface $eventDispatcher = null, MessageBusInterface $bus = null)
+    {
+        $this->timezone = \is_string($timezone) ? new \DateTimeZone($timezone) : $timezone;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->bus = $bus;
+
+        if (0 !== \count($pools)) {
+            $this->pools = $pools;
+        } else {
+            $this->createDefaultPool();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function schedule(TaskInterface $task): void
+    {
+        $task->setMultiples([
+            'priority' => null !== $task->get('priority') ? $task->get('priority') : Priority::NORMAL,
+            'scheduled_at' => new \DateTimeImmutable(),
+            'timezone' => $this->timezone,
+        ]);
+
+        $this->updateDefaultPool($task);
+
+        if (null === $this->bus) {
+            $this->tasks[$task->getName()] = $task;
+            $this->dispatchEvent(new TaskScheduledEvent($task));
+
+            return;
+        }
+
+        $this->bus->dispatch(new TaskMessage($task));
+        $this->dispatchEvent(new TaskScheduledEvent($task));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDueTasks(): TaskListInterface
+    {
+        return $this->getTasks()->filter(function (TaskInterface $task): bool {
+            return CronExpression::factory($task->get('expression'))->isDue(new \DateTimeImmutable());
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTimezone(): \DateTimeZone
+    {
+        return $this->timezone;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTasks(): TaskListInterface
+    {
+        return new TaskList($this->tasks);
+    }
+
+    public function getTasksByPool(string $poolName): TaskListInterface
+    {
+        return $this->getTasks()->filter(function (TaskInterface $task) use ($poolName): bool {
+            return $poolName === $task->get('pool');
+        });
+    }
+
+    public function getPool(string $poolName): TaskPoolInterface
+    {
+        return new TaskPool($poolName, $this->getTasks()->filter(function (TaskInterface $task) use ($poolName): bool {
+            return $poolName === $task->get('pool');
+        }));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPools(): array
```

Comment:
This is not used... I'm not sure to undersatnd the purpose of task pools

Reply:
For now, the pools are a draft, the idea is to create a task "group" which allow to group the tasks linked to something specific, think about task which are linked to "user creation" for example, you may want to define 2-3 tasks: 

- post-registration email
- same as first but 24H later if the account is not activated
- Welcome emails when the account is validated

Using a pool can be useful as we can regroup every tasks using the id of the user for example and schedule it every time a new user is created.

Reply:
For clarity, I've removed the idea of pools, the `TaskList` is enough 🙂